### PR TITLE
feat: add support for DDL and INSERT/DELETE/UPDATE operations

### PR DESCRIPTION
**Pull request by curino**
In collaboration with Jesus Camacho Rodriguez

### CODE REVIEW

**Comment by curino** [comment]
Cleaned up git commit log (I spend most of my time fighting this linter!).

**Comment by curino** [comment]
@jacques-n and others, how does this version look?

**Comment by jacques-n** [comment]
Hey @curino , this is on my list to review. Will try to get to it soon. 

One thing I still don't understand is the output item in writerel. Can you give a concrete example in pseudo code/description?

**Comment by curino** [comment]
Thanks @jacques-n!   

Most DBMS return information about how the insert/delete/update operation went. These takes few forms:

1. No output (default)
2. A simple count of how many tuples were affected by the operation
3. Returning the set of tuples deleted/inserted
4. Returning the before (or after, or both) image of an update
5. An arbitrary computation of the affected tuples

What I learned reading few doc pages from various systems is that this is to simplify the common case where apps want to be able to double check how an operation went (e.g., validate that roughly the right number of tuples were affected) and potentially abort/compensate if things were off. This would otherwise required temp tables and complex multi-table transactions, so it is a rather appreciated/used feature.

Including an `output` term allows us to capture any of the above scenarios in a compact way. Basically we are telling the systems to use `input` query to determin which tuples to change and `output` to report on what was done per user required semantics. In some systems there are special keywords the `output` query can refer to such as  BEFORE and AFTER (think of those as tables capturing the before and after image of the affected tuples). I think it is ok to expect that the system generating the substrait plan produces valid pairs of `input` and `output` Rels (some validation is possible, but not very deep). 

I considered to simplify this and hardcode the "# of affected tuples" semantics as a return value, but this does seem to miss what I read was a rather common use case.   I hope this helps clarify the rationale (happy to chat in the community call if it is easier).

Thanks,
Carlo

**Comment by curino** [comment]
Per conversation in latest community meeting, I modified the `output` mechanisms to leverage a spool (via references) for the BEFORE image, and a simple enum `NO_OUTPUT` or `MODIFIED_TUPLES` to control the operator output, that we can then process (e.g., with a `COUNT`) to shape the output of AFTER image changes. 

I also realized that in the original docs there were references to a bunch of other things (rotations, etc.) that are not currently represented (or at least I am not clear on what they mean). I would propose we converge on this patch, and then do further iterations to reflects further/fancier things (e.g., we left constraints and indexes out of table definitions as well).

**Comment by curino** [comment]
@jacques-n ping...

File: proto/substrait/algebra.proto [hunk 1]
```diff
@@ -364,6 +364,86 @@ message Rel {
   }
 }
 
+// A base table for writing. The list of string is used to represent namespacing (e.g., mydb.mytable).
+// This assumes shared catalog between systems exchanging a message.
+// it also includes a base schema, and default values
+message NamedTableWrite {
+  repeated string names = 1;
+  // The columns that will be modified (representing after-image of a schema change)
+  NamedStruct base_schema = 2;
```
[the remark this answers is not in the document]
Comment (jacques-n) [reply proto/substrait/algebra.proto:373]: We should also probably have an option around defaults: do we only apply for missing columns or do we also apply for null values. (The latter is a lot more expensive...)

File: proto/substrait/algebra.proto [hunk 2]
```diff
@@ -364,6 +364,86 @@ message Rel {
...
+  Rel input = 4;
+
+  enum DdlOp {
+    DDL_OP_UNSPECIFIED = 0;
+    DDL_OP_CREATE_TABLE = 1;
+    DDL_OP_DROP_TABLE = 2;
+    DDL_OP_ALTER_TABLE = 3;
+
+    DDL_OP_CREATE_VIEW = 4;
+    DDL_OP_CREATE_OR_REPLACE = 5;
```
[the remark this answers is not in the document]
Comment (curino) [reply proto/substrait/algebra.proto:402]: We could have the DDL_OP being TABLE or VIEW and then have another field capturing CREATE, ALTER, DROP (where VIEW  + ALTER is equivalent to the CREATE_OR_REPLACE we have now)

File: proto/substrait/algebra.proto [hunk 3]
```diff
@@ -364,6 +364,86 @@ message Rel {
...
+    DDL_OP_DROP_VIEW = 6;
+  }
+}
+
+// The operator that modifies the content of a database
+message WriteRel {
+  // Definition of which type of write operation is to be performed
+  oneof write_type {
+    NamedTableWrite named_table = 1;
+    ReadRel.LocalFiles local_files = 2;
```
[the remark this answers is not in the document]
Comment (curino) [reply proto/substrait/algebra.proto:412]: We could include tables only for now, but eventually we want `WriteRel` to write to file as well I think.

File: proto/substrait/algebra.proto [hunk 4]
```diff
@@ -364,6 +364,86 @@ message Rel {
...
+  DdlOp op = 3;
+
+  // The body of the CREATE VIEW / CTAS
+  Rel input = 4;
+
+  enum DdlOp {
+    DDL_OP_UNSPECIFIED = 0;
+    DDL_OP_CREATE_TABLE = 1;
+    DDL_OP_DROP_TABLE = 2;
+    DDL_OP_ALTER_TABLE = 3;
```
[the remark this answers is not in the document]
Comment (curino) [reply proto/substrait/algebra.proto:399]: If we keep the `write_type` it contains a `NamedTableWrite`, which includes all we need for CREATE or ALTER TABLE I think (table name, column names, column types, default values).

File: proto/substrait/algebra.proto [hunk 5]
```diff
@@ -364,6 +364,86 @@ message Rel {
   }
 }
 
+// A base table for writing. The list of string is used to represent namespacing (e.g., mydb.mytable).
+// This assumes shared catalog between systems exchanging a message.
+// it also includes a base schema, and default values
+message NamedTableWrite {
```
[the remark this answers is not in the document]
Comment (curino) [reply proto/substrait/algebra.proto:370]: I think this prevents us to do all `CREATE' and 'ALTER' operators.

File: proto/substrait/algebra.proto (hunk 1 above)
[the remark this answers is not in the document]
Comment (curino) [reply proto/substrait/algebra.proto:373]: The thinking is as follows: 

1. `names` is simply a qualified table name (like mydb.table). I am copying the pattern from `ReadRel.NamedTable`. Maybe we should call it `qualified_table_name` or something like that.
2. `base_schema` describes the entire schema we want in output (after an ALTER or as a result of CREATE or INSERt operation etc.)
3. The content is positionally (as usual) mapped to `input`. For example if you are doing UPDATE TABLE foo WHERE col1=3 WHERE cond, we would have in the `input` something like SELECT "3", col2, col3... FROM foo WHERE cond. 

In hindsight we might be able to skip the `defaults` field altogether and use the `input` for that (using a set of constants if we are issuing a `CREATE TABLE` for example).

**Comment by curino** [comment]
Thanks @jacques-n  for the live-chat. I have addressed all we discussed, and updated some of the docs accordingly.

**Review by jacques-n — APPROVED** [review]
One little leftover that should be cleaned but otherwise looking good. I'll fix and merge assuming tests pass.

**Review by jacques-n — APPROVED** [review]

**Comment by curino** [comment]
Thanks @jacques-n and @jcamachor for the help shaping this.

**Comment by jvanstraten** [comment]
I'm a bit late to the party now (was on vacation, sorry), but I can't help but notice that `oneof rel_type` wasn't updated. AFAICT the new messages are entirely unreachable from `Plan`. Is that intentional at this time? I haven't read any of this in detail yet though, so I might have missed something.

**Pull request merged**